nixos/*: deprecate docbook option docs

pennae opened the pull request:
###### Description of changes

deprecate docbook option docs and functions for creating docbook option docs. this very explicitly does not introduce warnings that could break a build to not break any systems when the 22.11 rolls around, and for 23.05 we'll want to remove docbook support for options entirely (without another warning).

a couple new module have been merged since the last conversion PR was started and a few options weren't converted even then, so we have to handle a few deprecation messages immediately as well.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pennae committed:
lib/options: deprecate docbook text and literalDocBook

deprecate literalDocBook by adding a warning (that will not fire yet) to
its uses and other docbook literal strings by adding optional warning
message to mergeJSON.

pennae committed:
nixos/*: fix docbook deprecation notices

mostly no rendering changes except in buildkite, which used markdown
where docbook was expected without marking up its markdown.

edolstra commented:
What is the advantage of `literalExpression` over `literalExample`? They seems to do the same thing, and `literalExample` has a more accurate name.

dasJ commented on nixos/modules/programs/rust-motd.nix:
```diff
@@ -7,7 +7,7 @@ let
 in {
   options.programs.rust-motd = {
-    enable = mkEnableOption "rust-motd";
+    enable = mkEnableOption (lib.mdDoc "rust-motd");
```
Not sure about this (and future `mkEnableOption` calls): Is this correct?
Because looking at the implementation of `mkEnableOption`, it seems like that would try to embed an attrset into a string.

dasJ replied:
To be more clear: I think this would result in:
```
lib.mkOption {
  type = lib.types.bool;
  default = false;
  example = true;
  description = "Whether to enable { _type = "mdDoc"; text = "rust-motd"; }.";
};
```

dasJ reviewed (approved):
nix looks good, changelog and docs look good.
Have only skimmed the lib.mdDoc insertions but they looked good from what I can tell.
Have not reviewed the Python code

infinisil commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -118,6 +118,8 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
 
 - The `isPowerPC` predicate, found on `platform` attrsets (`hostPlatform`, `buildPlatform`, `targetPlatform`, etc) has been removed in order to reduce confusion.  The predicate was was defined such that it matches only the 32-bit big-endian members of the POWER/PowerPC family, despite having a name which would imply a broader set of systems.  If you were using this predicate, you can replace `foo ...
 
+- `literalExample`, which has been deprecated since 21.11, has been removed.
```
This is too fast of a removal. I think the warning should be there at least for like 2 years. And it shouldn't be removed straight away either, a `throw` should be used after the warning has been there long enough.

roberth replied:
> keeping the _warning_ around for another year seems a bit long.

Can we put users first? An alias has basically 0 cost.
Why not keep it around so that even the slowest moving users know what happened?

infinisil commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -179,6 +181,8 @@ Use `configure.packages` instead.
 
 - memtest86+ was updated from 5.00-coreboot-002 to 6.00-beta2. It is now the upstream version from https://www.memtest.org/, as coreboot's fork is no longer available.
 
+- Option descriptions, examples, and defaults writting in DocBook are now deprecated and will generate informational messages during the documentation build. Using CommonMark is preferred and will become the default in 23.05.
```
This is also too fast. There needs to be a reasonable period for third-party tools to easily support multiple nixpkgs releases without causing warnings. This was previously a problem with iirc home-manager, where a deprecation warning was unavoidable due to older nixpkgs still needing support.

infinisil replied:
> anything that wants to support both can either go the trivial route with a let mdDoc = lib.mdDoc or lib.id; in prefix and using documentation that works with both renderers, or it can have two texts and switch between them based on pkgs.version

Both of those are far from ideal and require a lot of work. Third-party tools should also have a chance for a smooth transition with an appropriate deprecation period. Merging this PR as-is would require home-manager, nix-darwin (and anything else rendering docs using the module system) to effectively immediately update all their docbook to markdown, support for which was only introduced [a couple months ago](https://github.com/NixOS/nixpkgs/pull/176146).

This is how I think the deprecation strategy should look like:
- 22.11: Introduce support for markdown docs (https://github.com/NixOS/nixpkgs/pull/176146). Mention in the release notes that support for docbook will be deprecated in the next release and removed in the one after that. No warning for docbook yet. This allows third-party tools to read release notes and prepare for the next release, such that they won't cause any warnings for users.
- 23.05: Deprecate docbook support by adding a warning when `mdDoc` isn't used. The warning mentions that docbook support will be removed in the next released. This pushes everybody to use `mdDoc`, explicitly confirming that they have transitioned their code to markdown.
- 23.11: Remove docbook support either by using markdown implicitly for unannotated strings. This is only safe assuming everybody did the transition to 23.05, if somebody updates from older releases, they'd never get the warning to move to `mdDoc`, but considering the alternative of throwing an error for unannotated strings, that's a worthwhile tradeoff imo.
- (future): Slowly deprecate `mdDoc` with a warning, then an error, then a complete removal


So essentially there's two deprecation phases: An initial silent one for third-party tools to prepare a smooth NixOS/nixpkgs upgrade for their users, and a noisy one (warning) for end users. And if a third-party tool misses the first phase of deprecation, they will get complaints by their users during the second one.

infinisil replied:
Found the issue I was thinking of that shows how this problem happened in the past: https://github.com/NixOS/nixpkgs/pull/163451#event-6275878451

roberth commented on lib/options.nix:
```diff
@@ -275,16 +275,16 @@ rec {
   literalDocBook = text:
     if ! isString text then throw "literalDocBook expects a string."
-    else { _type = "literalDocBook"; inherit text; };
+    else
+      lib.warn
+        "literalDocBook is deprecated, use literalMD instead"
+        { _type = "literalDocBook"; inherit text; };
```
This will annoy third party module maintainers.
They'll receive issues and PRs about the warnings, while they need to keep using docbook until 22.05 is EOL.
You could prepare the deprecation though, with

```suggestion
    else
      lib.warnIf (lib.isInOldestRelease 2211)
        "literalDocBook is deprecated, use literalMD instead"
        { _type = "literalDocBook"; inherit text; };
```

roberth replied:
They can't switch because they need to keep using docbook until 22.05 is EOL.

Delaying the warnings until the moment they can switch solves the problem for them.

For reference, this problem was described in https://github.com/NixOS/nixpkgs/issues/157667.
Machinery to automate deprecations was added in https://github.com/NixOS/nixpkgs/pull/163451.

roberth commented:
> add a temporary option to be removed together with docbook, eg `documentation.nixos.options.renderFromMD` that defaults to `false` and allows opting in to the fully docbook-less option docs world as soon as the option comes around.

sgtm

> the unstabe manual should be built with both values for this option for time being, deprecation notices in the build should count as errors, and differences between the two build should count as errors as well.

Yes, for Nixpkgs' own CI we can and should be strict. I can think of two ways we should probably both implement
 - add `renderFromMD = true` to the `nixosTests`; perhaps even as a channel blocker(?)
 - configure the github action that builds the manual to set the option, or duplicate the action

These changes won't affect users, but keep nixpkgs prepared for the final change.

pennae committed:
workflows: add check for docbook/md manual equality

we want to make sure that rendering the manual from markdown without
going through docbook produces (semantically) the same output as with
going through docbook. to ensure this we'll build the manual twice, run
each manual through html-tidy to generate a normalized form and diff
the normalized forms. we don't want to compare raw output because that
exposes us to a lot of whitespace we'd have to reproduce exactly in the
MD render.

this check may be relaxed even further in the future, but hopefully not
by much.

pennae committed:
workflows: check that nixos manual does not use docbook option docs

the nixos manual should not use docbook for module option documentation,
only markdown, to make future transition to a markdown-only world easier
and less painful. this check will ensure that all options
documentation (even plain text that would not be interpreted specially
by neither markdown nor docbook) is declared as being markdown.

roberth commented on .github/workflows/manual-nixos.yml:
```diff
@@ -29,3 +29,18 @@ jobs:
+      - name: Compare DocBook and MD manuals
+        run: |
+          export NIX_PATH=nixpkgs=$(pwd)
+          nix-build \
+            --option restrict-eval true \
+            -o docbook nixos/release.nix \
+            -A manual.x86_64-linux
+          nix-build \
+            --option restrict-eval true \
+            --arg configuration '{ documentation.nixos.options.renderFromMD = true; }'
+            -o md nixos/release.nix \
+            -A manual.x86_64-linux
+          .github/workflows/compare-manuals.sh \
+            docbook/share/doc/nixos/options.html \
+            md/share/doc/nixos/options.html
```
Could you add explanatory messages to these commands when they fail?
That way, the average contributor will know what to do in either of the three circumstances.
Would a failure in the last command only be caused by a programming error in the python script?

zowoq commented on .github/workflows/manual-nixos.yml:
```diff
@@ -29,3 +29,25 @@ jobs:
           signingKey: '${{ secrets.CACHIX_SIGNING_KEY }}'
       - name: Building NixOS manual
         run: NIX_PATH=nixpkgs=$(pwd) nix-build --option restrict-eval true nixos/release.nix -A manual.x86_64-linux
+      - name: Compare DocBook and MD manuals
```
I don't think this should be added here. 

This is a contributor facing check avoid the manual failing to build and then blocking nixos channels. 

I don't see that comparing docbook and md is relevant to this check, or why it even needs to be a github action?

zowoq replied:
> neither of those need to be done in an action ... and they can still be merged regardless if they're important enough

Then I don't understand why it's being added here. This is intentionally a very simple check, i don't want people to ignore it.

> for either problem to be resolved before they are merged 

What problems? I don't see how anything from this is actionable by a contributor.  "open a issue" means it shouldn't be in this check.

zowoq replied:
> what do you propose we do instead

I don't understand why it needs to be a PR check?

zowoq replied:
> it checks for consistency of the nixpkgs documentation with the proposed policy ("don't use docbook in nixpkgs"), that seems like a pretty reasonable thing to put in an action.

Yes, this seems fine as it would be something that would be problematic in PRs and would be better resolved in the PR before merge.

> the other check ("docbook manual and md manual render the same") extends that to make sure the two options to produce the manual don't produce wildly different results

This doesn't seem like something a casual contributor should ever need to see.

> is there a better place to put that? a nixos test could also check that, but if it's not a channel blocker it'll be routinely forgotten and thus not very useful.

Could have an action on cron that posts a comment like we have for merge branches?

infinisil commented on nixos/lib/make-options-doc/mergeJSON.py:
```diff
@@ -241,9 +250,32 @@ def optionIs(option: Dict[str, Any], key: str, typ: str) -> bool:
 hasWarnings = False
 for (k, v) in options.items():
+    if warnOnDocbook:
+        if isinstance(v.value.get('description', {}), str):
+            hasWarnings = True
+            print(
+                f"\x1b[1;31m{severity}: option {v.name} description uses DocBook\x1b[0m",
+                file=sys.stderr)
```
Why is `.description` not checked for `literalDocBook` with `is_docbook`?

infinisil replied:
Oh right, it's the `description`, which is always literal

infinisil commented on nixos/lib/make-options-doc/default.nix:
```diff
@@ -34,6 +34,10 @@
 # instead of printing warnings for eg options with missing descriptions (which may be lost
 # by nix build unless -L is given), emit errors instead and fail the build
 , warningsAreErrors ? true
+# render directly from markdown instead of going through a docbook intermediate stage.
+# currently this only forbids using docbook in option docs, it does not yet render
+# directly from markdown.
+, renderFromMD ? false
```
Why is this flag called `renderFromMD` when it doesn't do that? If it just warns for docbook shouldn't it be called something like `allowDocbook = true/"warn"/"error"`?

infinisil replied:
It is very misleading though. How about having such a proposed `allowDocbook` and if we later allow markdown rendering, we can introduce another option `renderBackend = "docbook"/"markdown"`. Then if "markdown" is chosen, `allowDocbook` is forcibly set to "error" or `null` (indicating `docbook` is not even detected anymore, so the option doesn't make sense)

infinisil reviewed (changes requested):
Almost looking good to me!

infinisil commented:
I discussed this with @pennae on Matrix [here](https://matrix.to/#/!kjdutkOsheZdjqYmqp:nixos.org/$QozPn7j3i42kZ0bz0Y75sAl3AIoDQdpce7IIB_AP77Q), we concluded that `allowDocbook` was okay, for now both `true` and `false` values use the docbook render pipeline, but when markdown pipeline support is added, setting `allowDocbook = false` can use that one instead to speed up the doc build.

infinisil commented on nixos/modules/misc/documentation.nix:
```diff
@@ -255,6 +255,23 @@ in
+          The `false` setting for this option is not yet fully supported. While it
+          should work fine and produce the same output as the previous toolchain
+          using DocBook it may not work in all circumstances. It is always permissible
+          to set this to `true` when it was set to `false` previously.
```
Wait what does this mean? "It is always permissible to set this to `true` when it was set to `false` previously."?

infinisil replied:
I see, how about rewording this a bit then, something like "Whether markdown option documentation is allowed is independent of this option."

vcunat commented:
@pennae: this PR broke nixpkgs manual: https://hydra.nixos.org/build/190038194

zowoq commented:
> this PR broke nixpkgs manual: https://hydra.nixos.org/build/190038194

https://github.com/NixOS/nixpkgs/pull/190751

jtojnar commented on nixos/doc/manual/development/option-declarations.section.md:
```diff
@@ -44,26 +44,23 @@ The function `mkOption` accepts the following arguments.
+    https://nixos.org/nixpkgs/manual/#sec-contributing-markup) format, that will be
     included in the NixOS manual. During the migration process from DocBook
-    to CommonMark the description may also be written in CommonMark, but has
-    to be wrapped in `lib.mdDoc` to differentiate it from DocBook. See
```
Looks like `lib.mdDoc` is no longer documented in the NixOS manual: https://nixos.org/manual/nixos/unstable/#sec-option-declarations

This leads to confusion: https://github.com/NixOS/nixpkgs/pull/200724

amjoseph-nixpkgs commented:
Hi, this error is triggering on a package (`iay`) that my PR does not touch, and the error message gives no indication as to the location of the docbook it accuses me of using:

https://github.com/NixOS/nixpkgs/actions/runs/3846394423/jobs/6551643904
